### PR TITLE
[Annotation] Don't rotate an annotation when it has the NoRotate flag

### LINKED ISSUE #7631: comments on PDF files in landscape orientation are rotated and in wrong position
There is a problem in pdf.js when displaying comments on documents in landscape orientation.  The comment icon and popup text are shown as rotated 90 degree.  Documents in portrait orientation don't have this problem.

Link to PDF file (or attach file here):
Please see the attached pdf_landscape_adobe.pdf
[pdf_landscape_adobe.pdf](https://github.com/mozilla/pdf.js/files/470518/pdf_landscape_adobe.pdf)

Configuration:
- Web browser and its version:  Google Chrome Version 52.0.2743.116 m
- Operating system and its version:  Windows 7 Professional Service Pack 1
- PDF.js version: 1.4.20
- Is an extension: No.

Steps to reproduce the problem:
1. Create a pdf document in landscape orientation
2. Added a comment and an annotation (for comparation purpose) with Adobe Reader.  Both comment and annotation look good in Adobe Reader.  See the attached pdf_landscape_adobe.pdf.
3. Go to pdf.js demo site http://mozilla.github.io/pdf.js/web/viewer.html and open the above pdf file.  
4. You can see the annotation is displayed correctly, but the comment has problems.
- The icon is rotated 90 degree.
- The popup text (once you click on the icon) is rotated 90 degree.
- The popup text is not in the right position.

What is the expected behavior? (add screenshot)
The comment should be displayed correctly as annotation.  See screenshot 
pdf_landscape_expected
pdf_landscape_bug

pdf_landscape_expected.png. 

What went wrong? (add screenshot)
The following problems were observed.  See screenshot pdf_landscape_bug.png.
- The icon is rotated 90 degree.
- The popup text (once you click on the icon) is rotated 90 degree.
- The popup text is not in the right position.

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):
http://mozilla.github.io/pdf.js/web/viewer.html

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

File: src/core/annotation.js [hunk 1]
```diff
@@ -3433,6 +3435,10 @@ class TextAnnotation extends MarkupAnnotation {
 
     super(params);
 
+    // No rotation for Text (see 12.5.6.4).
+    this.noRotate = true;
```
Comment (Snuffleupagus) [inline src/core/annotation.js:3439]: I'm guessing that you actually meant the following here?
```suggestion
    this.data.noRotate = true;
```

File: src/core/annotation.js [hunk 2]
```diff
@@ -3433,6 +3435,10 @@ class TextAnnotation extends MarkupAnnotation {
...
+    this.data.hasOwnCanvas = true;
```
Comment (Snuffleupagus) [inline src/core/annotation.js:3440]: For overall consistency, given all classes below, how about the following here as well?
```suggestion
    this.data.hasOwnCanvas = this.data.noRotate;
```

File: web/annotation_layer_builder.css [hunk 3]
```diff
@@ -53,6 +53,21 @@
...
 
+.annotationLayer[data-main-rotation="270"] .norotate {
+  transform: rotate(90deg) translateY(-100%);
+}
+.annotationLayer[data-main-rotation="180"] .norotate {
+  transform: rotate(180deg) translate(-100%, -100%);
+}
+.annotationLayer[data-main-rotation="90"] .norotate {
+  transform: rotate(270deg) translateX(-100%);
+}
```
Comment (Snuffleupagus) [inline web/annotation_layer_builder.css:64]: Can we please re-order these rules as follows?
```suggestion
.annotationLayer[data-main-rotation="90"] .norotate {
  transform: rotate(270deg) translateX(-100%);
}
.annotationLayer[data-main-rotation="180"] .norotate {
  transform: rotate(180deg) translate(-100%, -100%);
}
.annotationLayer[data-main-rotation="270"] .norotate {
  transform: rotate(90deg) translateY(-100%);
}
```

**Review by Snuffleupagus — DISMISSED** [review]
r=me, with comments addressed and all tests passing (or improving); thank you!

File: web/annotation_layer_builder.css (hunk 3 above)
Comment (calixteman) [reply web/annotation_layer_builder.css:64]: Oh yes I forgot that... copy/paste side effect.

File: web/annotation_layer_builder.css [hunk 4]
```diff
@@ -237,6 +249,18 @@
   width: 100%;
 }
 
+.annotationLayer .fileAttachmentAnnotation .popupTriggerArea {
+  position: absolute;
+  top: 0;
+  left: 0;
+}
+
+.annotationLayer .fileAttachmentAnnotation .popupWrapper {
```
Comment (Snuffleupagus) [inline web/annotation_layer_builder.css:258]: This looks like unnecessary duplication, can't we just do the following instead?
```suggestion
.annotationLayer .fileAttachmentAnnotation .popupTriggerArea,
.annotationLayer .fileAttachmentAnnotation .popupWrapper {
```

Also, what's "special" about the FileAttachmentAnnotation that required these changes?

File: web/annotation_layer_builder.css [hunk 5]
```diff
@@ -237,6 +250,10 @@
   width: 100%;
 }
 
+.annotationLayer .fileAttachmentAnnotation .popupTriggerArea {
+  position: absolute;
+}
+
```
Comment (Snuffleupagus) [inline web/annotation_layer_builder.css:256]: In addition to the FileAttachmentAnnotation case, the `popupTriggerArea`-class may also be used as a fallback (see below).
So, should we perhaps just move the `position: absolute;` into the CSS rule *just above* instead?

https://github.com/mozilla/pdf.js/blob/652a6df62bdc6a1765f64714d96b9454433f847f/src/display/annotation_layer.js#L454-L459

Comment (calixteman) [reply web/annotation_layer_builder.css:256]: I tried that and some popups are moving in the reftests. From my pov it doesn't matter that much but I wanted to minimize the change. I'll push and updated patch and we'll see what's the best.

Comment (Snuffleupagus) [reply web/annotation_layer_builder.css]: Thanks for testing that!

Looking at the results, we probably want limit this to just the FileAttachmentAnnotation to reduce "unrelated" changes in this patch.

**Review by Snuffleupagus — APPROVED** [review]
r=me, thank you!

**Pull request merged**